aten/src/ATen/native/NamedTensor.cpp

```diff
@@ -16,5 +16,141 @@ Tensor set_names(const Tensor& self, optional<DimnameList> names) {
   return result;
 }
 
+static void report_moving_unnamed_dim_error(
+    DimnameList names, DimnameList other, bool is_aligning_two_tensors) {
+  if (is_aligning_two_tensors) {
+    TORCH_CHECK(false,
+        "Aligning Tensor", names, " and Tensor", other,
+        " would change the absolute position from the right of an unnamed dimension. ",
+        "Please name unnamed dimensions to avoid ambiguity.");
+  } else {
+    TORCH_CHECK(false,
+        "Aligning Tensor", names, " to `names` ", other,
+        " would change the absolute position from the right of an unnamed dimension. ",
+        "Please name unnamed dimensions to avoid ambiguity.");
+  }
+}
+
+static void report_not_a_subsequence_error(
+    DimnameList names, DimnameList other, bool is_aligning_two_tensors) {
+  if (is_aligning_two_tensors) {
+    auto shorter = names.size() > other.size() ? other : names;
+    auto longer = names.size() > other.size() ? names : other;
+    TORCH_CHECK(false,
+        "Could not align Tensor", shorter, " and Tensor", longer,
+        " because ", shorter, " is not a subsequence of ", longer, ". ");
+  } else {
+    TORCH_CHECK(false,
+        "Could not align Tensor", names, " to `names` ", other,
+        " because ", names, " is not a subsequence of `names`.");
+  }
+}
+
+static std::vector<Dimname> default_names(size_t len) {
+  return { len, Dimname::wildcard() };
+}
+
+// Let tensor `t` have size `tensor_sizes` and `tensor_names`.
+// This helper function computes the resulting size of `t` after aligning it
+// to `aligned_names`. Enforces the alignment rules in Note [Alignment rules].
+static std::vector<int64_t> aligned_size(
+    IntArrayRef tensor_sizes,
+    DimnameList tensor_names,
+    DimnameList aligned_names,
+    bool is_aligning_two_tensors) {
+  std::vector<int64_t> expanded_sizes(aligned_names.size(), 1);
+  ptrdiff_t dim = (ptrdiff_t)tensor_sizes.size() - 1;
+  ptrdiff_t idx = (ptrdiff_t)aligned_names.size() - 1;
```

Comment:
If I understand this right, you're explicitly casting to a signed type so that you can check `>= 0` below? I guess that's fine, but seems a bit inelegant

Reply:
Yes